File: Assets/Lua/ecs/systems/animation_command_system.lua
---
--- Created by echo.
--- DateTime: 2025/12/28 22:47
---
-- AnimationCommandSystem.lua
---@class AnimationCommandSystem
AnimationCommandSystem = {}
AnimationCommandSystem.__index = AnimationCommandSystem

---@param world World
---@param dt number
function AnimationCommandSystem:update(world, dt)
    ---@type table<integer, AnimationComponent>
    local animations = world:GetComponentOfType(_G.ComponentRegistry.Animation)
    ---@type table<integer, AnimationCommandComponent>
    local animation_commands = world:GetComponentOfType(_G.ComponentRegistry.AnimationCommand)

    for eid, command in pairs(animation_commands) do
        local anim = animations[eid]
        if not anim then
            goto continue
        end

        if command.play then
            local needRestart =
                command.forceRestart
                or anim.state ~= command.play

            if needRestart then
                anim.state = command.play -- ★关键
                anim.clipId = command.play
                anim.frame = 1
                anim.time = 0
                anim.playing = true
            end
        end

        if command.stop then
            anim.playing = false
        end

        world:RemoveComponent(eid, _G.ComponentRegistry.AnimationCommand)
        ::continue::
    end
end

return AnimationCommandSystem

File: Assets/Scripts/ECS/Systems/AnimationCommandSystem.cs
using System.Collections.Generic;
using ECS.Core;

namespace ECS.Systems
{
    /// <summary>
    /// 动画指令系统
    /// 职责: 执行动画指令
    /// </summary>
    public class AnimationCommandSystem : SystemBase
    {
        public override void Update(World world, float deltaTime)
        {
            var toRemove = new List<int>();

            foreach (var (entity, command) in world.GetComponents<AnimationCommandComponent>())
            {
                if (!world.HasComponent<AnimationComponent>(entity))
                {
                    toRemove.Add(entity);
                    continue;
                }

                var animation = world.GetComponent<AnimationComponent>(entity);

                switch (command.command)
                {
                    case "play":
                        animation.AnimName = command.anim_name;
                        animation.Frame = 1;
                        animation.Time = 0f;
                        animation.Playing = true;
                        break;
                    case "stop":
                        animation.Playing = false;
                        break;
                }

                // 执行回调
                command.callback?.Invoke();

                toRemove.Add(entity);
            }

            // 移除已执行的指令
            foreach (var entity in toRemove)
            {
                world.RemoveComponent<AnimationCommandComponent>(entity);
            }
        }
    }
}